Add sponsors

bastnsa opened the pull request:
## what does this PR do?

- [ ] Adds a new article
- [x] Fixes a bug / Implements a feature (please describe below)

Implements a funny parody sponsorship system. 
  1. Sponsors are stored in sponsors.json with sponsor properties (title, author handle, display name, image path, and alt text).
  2. A SponsorBanner component renders sponsors in two positions - on home/listing pages and on article pages.
  3. Placement details:
    - On listing pages: Shows 2 sponsors at calculated positions (roughly in thirds)
    - On article pages: Places a sponsor banner after content
    - Uses some randomization to ensure consistent sponsor display (the more sponsors the better)
  4. Added a CLI script (new-sponsor.ts) that works very similar to (new-article.ts):
    - Collects sponsor details interactively
    - Validates and processes images
    - Updates the sponsor json file
  5. Sponsor images are stored in the public directory with standardized slug naming.

## justification

- [ ] If you are adding a new article, why is it funny? (please explain below)
- [x] If you are implementing a feature, why is it important? (please explain below)

My reasons:
    - Creates opportunity for additional jokes and commentary
    - Adds visual humor that complements the written articles
    - More variety in the type of content (quicker and easier to contribute)

## checklist

- [x] I have read the [contributing guide](https://github.com/TehcCringe/TehcCringe/blob/master/docs/contributing.md)
- [x] I have read the [rules](https://github.com/TehcCringe/TehcCringe/blob/master/docs/rules.md)
- [x] I have read the [style guide](https://github.com/TehcCringe/TehcCringe/blob/master/docs/style-guide.md)

## would your mom approve of this PR?

- [x] Yes
- [ ] No

If you selected no, please explain why

IroncladDev commented on .eslintrc.json:
```diff
@@ -1,6 +1,7 @@
   "rules": {
-    "@next/next/no-img-element": "off"
+    "@next/next/no-img-element": "off",
+    "@typescript-eslint/no-require-imports": "off"
```
You shouldn't need to use `require` in an ESM environment unless you're trying to work around a circular dependency, which is extremely rare

IroncladDev commented on app/components/sponsor-banner.tsx:
A couple of nits on code style: 

1. I prefer using `export default function SponsorBanner` instead of `const SponsorBanner`
2. To make it easier to edit the file, please add the styled tailwind components **after** the main component near the end of the file

IroncladDev commented on app/lib/utils.ts:
```diff
@@ -5,3 +5,39 @@ export const slugify = (str: string) => {
+
+// Get sponsors for a specific page using randomization
+export const getSponsorsForPage = (pageKey: string, count: number = 1) => {
+  const sponsorsData = require("../../sponsors/sponsors.json")
```
Use `fs` or a dynamic import instead of `require` for this

IroncladDev commented on sponsors/sponsors.d.ts:
Please use `zod` to create a schema so the value of a `Sponsor` can be validated at runtime.

A `.d.ts` file isn't necessary. You can just use `.ts`

See `app/utils/articles.ts` for how I did this.

IroncladDev commented on sponsors/sponsors.json:
I would prefer to use file-based routing sort of how the `articles` directory works with each sponsor having its own directory with an image and a markdown/json file for info.

IroncladDev replied:
Yeah go ahead and copy the assets into `public/assets/` just like how articles do it, at build time

IroncladDev reviewed (changes requested):
Overall great PR. Requested some changes. Ping me once you get those done and I'll take another pass.

bastnsa commented:
> Overall great PR. Requested some changes. Ping me once you get those done and I'll take another pass.

Thanks for the feedback! I've addressed all the comments

IroncladDev commented on app/[page]/page.tsx:
```diff
@@ -151,6 +155,16 @@ export async function generateStaticParams() {
     chunks.push({ page: i.toString() })
   }
 
+  const sponsors = getAllSponsors()
```
copying the sponsor assets into the `assets` directory has nothing to do with generating static params. The `Home` function is a server component generated at build time so you can move this logic into there.

IroncladDev reviewed (changes requested):
Requested some final changes. After that, should be good to merge!

bastnsa commented:
> Requested some final changes. After that, should be good to merge!

Once again, I appreciate the detailed feedback, thanks!
